### PR TITLE
[otelarrowreceiver, otelarrowexporter] Add internal/grpcutil package

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** 
With utilities to encode and decode timeout in the syntax specified by gRPC.
This will be used by the otelarrow components to encode and decode timeout values.
In that sense, the new code could also be added into the pending internal/otelarrow package, see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33579

As this code is derived from gRPC-Go, some text is added in `NOTICE` according to the license. The code that this was derived from contained a TODO, which was largely addressed by eliminating very short timeouts from being encoded. This code will encode timeouts in milliseconds, seconds, hours, and days but it will not encode microsecond/nanosecond durations, these are rounded to `0m`.  The decoder will handle all durations that gRPC specifies, so that this logic can change in the future.

**Link to tracking Issue:** https://github.com/open-telemetry/otel-arrow/issues/227

**Testing:** Adds basic tests.

### CODE REVIEW

**Review by moh-osman3 — APPROVED** [review]

**Comment by jmacd** [comment]
This is part of https://github.com/open-telemetry/otel-arrow/issues/227.

We were encouraged to move development of the otel-arrow components into this repository, but we need our PRs to merge for this to continue. :-)

**Review by mwear — APPROVED** [review]
LGTM. It'd be nice if a maintainer could double check the contents of NOTICE.

**Comment by jmacd** [comment]
@mwear I would be willing to move this module under `internal/otelarrow` as well. It's the same owners. It's meant for the two OTel-Arrow components to use, but it could be useful outside OTel-Arrow.

**Comment by jmacd** [comment]
> It'd be nice if a maintainer could double check the contents of NOTICE.

For the record, the ASLv2 section 4(b) directs us to do this. There was already a NOTICE file concerning bits of gopsutil that were copied, and I just copied the pattern.

```
      (d) If the Work includes a "NOTICE" text file as part of its
          distribution, then any Derivative Works that You distribute must
          include a readable copy of the attribution notices contained
          within such NOTICE file, excluding those notices that do not
          pertain to any part of the Derivative Works, in at least one
          of the following places: within a NOTICE text file distributed
          as part of the Derivative Works; within the Source form or
          documentation, if provided along with the Derivative Works; or,
          within a display generated by the Derivative Works, if and
          wherever such third-party notices normally appear. The contents
          of the NOTICE file are for informational purposes only and
          do not modify the License. You may add Your own attribution
          notices within Derivative Works that You distribute, alongside
          or as an addendum to the NOTICE text from the Work, provided
          that such additional attribution notices cannot be construed
          as modifying the License.
```

**Review by atoulme — APPROVED** [review]
LGTM - my take is this is an internal package, it helps now, ok to get it in

**Comment by jmacd** [comment]
🥺 I think this could merge?

**Review by moh-osman3 — APPROVED** [review]

**Comment by jmacd** [comment]
Please merge this.

**Comment by jmacd** [comment]
I apologize for pushing this PR without apparently much cause. As we know, it is difficult to maintain multi-module PRs in this repository especially when introducing a new module. I could send a PR containing this PR combined with changes in otelarrowreceiver and one in otelarrowexporter. I could, but it would become harder to maintain and harder to review, so I've been waiting for this one to merge before sending more PRs. https://github.com/open-telemetry/otel-arrow/issues/227

File: internal/grpcutil/timeout.go
Comment (evan-bradley) [inline internal/grpcutil/timeout.go:1]: Is this all from `internal/transport/http_util.go`?

Section 4(b) of Apache 2.0 says:

>You must cause any modified files to carry prominent stating that You changed the files; and

Should we include a note here about that?

**Review by evan-bradley — APPROVED** [review]
Looks good to me overall, I don't have concerns with adding additional internal modules.

I'm not totally clear on the license requirements, but what's here seems sufficient to me. I'll merge this after the release is completed tomorrow if there aren't further comments.

File: internal/grpcutil/timeout.go
Comment (jmacd) [reply internal/grpcutil/timeout.go:1]: I've added a note in timeout.go describing the two locations that were copied from.

Comment (jmacd) [reply internal/grpcutil/timeout.go:1]:
```
// The EncodeTimeout function is forked and modified from the original
// https://github.com/grpc/grpc-go/blob/master/internal/grpcutil/encode_duration.go

// This DecodeTimeout function is forked and modified from the original
// https://github.com/grpc/grpc-go/blob/master/internal/transport/http_util.go
```

**Comment by mx-psi** [comment]
@evan-bradley is this good to merge?

File: internal/grpcutil/timeout.go
Comment (evan-bradley) [reply internal/grpcutil/timeout.go:1]: Perfect, thanks!

**Pull request merged**
